### ISSUE TITLE
OOP variance rules + FP

### ISSUE THREAD

**Issue by shanemikel** [palatable/lambda#1]
Here, http://jnape.com/the-perils-of-implementing-functor-in-java/, you assert that the OOP inheritance (variance) in Java is at odds with the FP functor.  I am rather fresh in my understanding of the Java and Haskell systems, but I would appreciate if you could elaborate on this issue.

I intend to contemplate this further... perhaps you could give me a head start.  Is there another variance rule that would permit the encoding of the functor abstraction to your satisfaction?  What does OCaml do?

**Comment by jnape** [comment]
Well I suppose this is what I get for not enabling comments on the blog then.

I will caveat my response by saying that regardless of how much time I spend writing this answer, in all likelihood I spent more time writing the blog post, so if I couldn't explain it well there, I'm dubious of my ability to do better here.

Let's start with the minimum complete definition for `Functor`, as it's parametrically defined in Haskell:

```haskell
class Functor f where
  fmap :: (a -> b) -> f a -> f b
```

We can say about `fmap` that it maps _covariantly_ from `Functor a` to `Functor b`, which is clear observationally, as we see that the arrows match between the function `a -> b` and the variance of `f a -> f b` (this is not always such a straightforward way to determine variance, but `Functor`'s simplicity obviates this). 

We also know from the type signature that f's [_kind_](https://wiki.haskell.org/Kind) is `* -> *`, meaning it is parametric (in this case, singly, or having only one type parameter).

In Haskell, we can trivially arrive at a single parametric type signature from any multiple parametric type constructor just by supplying type parameters until we arrive at the kind we're looking for. So, for instance, `Either a b` is dually parametric, so in order to be considered as a `Functor`, it needs to have one of its parameters (the first parameter, `a`) supplied.

Another example of a type constructor that is dually parametric is `(->) a b`: it is a function from `a` to `b`. So, how can we elect `(->)` into `Functor`? We need to supply the first type parameter, `a`, which in the case of `(->)`, is the argument to the function, leaving the only unknown parameter `b`, the return type. Now the question is, how do we implement `fmap`?

First let's ask, what does `fmap` mean in the context of `(->) a`? Let's start by specifying `fmap`'s type signature in terms of `(->) a`:

```haskell
instance Functor ((->) r) where
  fmap :: (a -> b) -> ((->) r) a -> ((->) r) b
```

...which, could be written more intuitively as:

```haskell
instance Functor ((->) r) where
  fmap :: (a -> b) -> (r -> a) -> (r -> b)
```

In English, we could read `fmap` as taking a function from `a` to `b`, and a function from `r` to `a`, and finally returning a function from `r` to `b`.

Which is, of course, function composition. This is also a useful reason for Haskell's composition operator [(.)](https://wiki.haskell.org/Function_composition) to expect the functions in the order it does: it takes first `f` and then `g`, producing a function `\x -> f(g(x))`. For this reason, compose in Haskell could simply be written as:

```haskell
(.) :: (b -> c) -> (a -> b) -> (a -> c)
(.) = fmap
```

So, you could make the case (which I do) that `(->) r` being a `Functor` proves the existence of composition. It would then seem reasonable to represent composition in terms of `fmap`, wouldn't it? 

Except that in many languages, you can't.

Let's revisit the salient point here which is that the reason this identity holds is that `fmap` expects the mapping function as the first argument and the `Functor a` as the second.

How would you represent this type signature in Java? You might ([I do](https://github.com/palatable/lambda/blob/master/src/main/java/com/jnape/palatable/lambda/functor/Functor.java)) represent it as:

```java
public interface Functor<A> {
    <B> Functor<B> fmap(MonadicFunction<? super A, ? extends B> fn);
}
```

The troubling bit comes as a result of how method invocation is semantically represented in Java, and indeed, most object-oriented languages: `type.method(args)`. Substituting our previous `fmap` example in Java, this becomes:

```java
functorA.fmap(functionFromAToB); // -> functorB
```

..which, when functorA is a _function_, reads as:

```java
fnAToB.fmap(fnBToC); // -> fnAToC
```

Which is still compose, but in the wrong order, `(g . f) = \x -> f(g(x))`, instead of Haskell's way, `(f . g) = \x -> f(g(x))`. This is why Java [calls this method "andThen"](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html#andThen-java.util.function.Function-) on its own `Function` interface. In fact, in Java it _must_ be this way, because the unknown type parameter here, `<B>`, must be unified at the method call site, which is too late for the `Functor<B>` to be supplied as the left-hand side term given Java's current dispatch rules and semantics. I find things like this to be deliciously ironic, but in a rather sinister way; hence the blog post.

As far as what OCaml does, I don't believe OCaml has a built in composition operator, which, depending on your perspective, is quite possibly a life-threatening amount of irony.

So, to sum it up, Haskell's `fmap` is a function tied to the `Functor` typeclass, but Haskell doesn't require the instance argument to be in a strict position, allowing the flexibility to write things like the compose operator in terms of `fmap`, whereas in Java, or languages that do require a strict position of their instance arguments, this cannot be done.

**Comment by jnape** [comment]
For what it's worth, there was some level of discussion about this on Lobsters following the post.

https://lobste.rs/s/4b159w/perils_implementing_functor_java
